src/transaction/transaction_manager.cpp

```diff
@@ -5,21 +5,40 @@
 
 namespace terrier::transaction {
 TransactionContext *TransactionManager::BeginTransaction() {
-  // This latch has to also protect addition of this transaction to the running transaction table. Otherwise,
-  // the thread might get scheduled out while other transactions commit, and the GC will deallocate their version
-  // chain which may be needed for this transaction, assuming that this transaction does not exist.
-  common::SharedLatch::ScopedSharedLatch guard(&commit_latch_);
-  timestamp_t start_time = time_++;
+  timestamp_t start_time;
+  {
+    // There is a three-way race that needs to be prevented.  Specifically, we
+    // cannot allow both a transaction to commit and the GC to poll for the
+    // oldest running transaction in between this transaction acquiring its
+    // begin timestamp and getting inserted into the current running
+    // transactions list.  Using the current running transactions latch
+    // prevents the GC from polling and stops the race.  This allows us to
+    // replace acquiring a shared instance of the commit latch with a
+    // read-only spin-latch and move the allocation out of a critical section.
+    common::SpinLatch::ScopedSpinLatch running_guard(&curr_running_txns_latch_);
```

Comment:
How do you plan to replace this with thread-local implementation of `curr_running_txns` in the future?

Reply:
Assuming the thread's context is passed in as a parameter (same as project 1), then you just replace this latch and the corresponding container with the latch and container in the thread context.